--// Compilation Date: Wed Sep 03 2025 14:25:01 GMT-0400 (Eastern Daylight Time)
return {
	["_locale"] = "en_crp",
	["_locale_display"] = "English (Corporate Speak)",
	["_locale_display_localized"] = "English (Corporate Speak)",
	["Apps.MainFrame.Config.Desc"] = "Modify your panel with ease.",
	["Apps.MainFrame.Config.Title"] = "Modifications",
	["Apps.MainFrame.Home.Desc"] = "Your go-to spot for quick resources.",
	["Apps.MainFrame.Home.Title"] = "Residence",
	["Apps.MainFrame.Template.Desc"] = "Your go-to spot for quick resources.",
	["Configuration.Admins.Container.Ranks.Content.Template.Info"] = "%s • %s • Incepted by @%s %s",
	["Configuration.Admins.Ranks.Protected"] = "Protected",
	["Configuration.Admins.Ranks.Unprotected"] = "Not protected",
	["Configuration.Admins.Ranks.HasApps"] = "%s Applications",
	["Configuration.Admins.Ranks.HasSuperAdmin"] = "Complete control",
	["Configuration.Admins.Container.Ranks.Header.TLabel"] = "Customer",
	["Configuration.Admins.EditRank.Apps.Content.Template.AppName"] = "Modifications",
	["Configuration.Admins.EditRank.Apps.Header.TLabel"] = "Applications & Modifications",
	["Configuration.Admins.EditRank.Header.Finish.HeaderLabel"] = "Complete",
	["Configuration.Admins.EditRank.Header.TLabel"] = "Modifying %s",
	["Configuration.Admins.EditRank.People.Content.GroupTemplate.Label"] = "Community",
	["Configuration.Admins.EditRank.People.Content.GroupTemplate._Name"] = "Community (%s followers)",
	["Configuration.Admins.EditRank.People.Content.PlayerTemplate.Label"] = "Visitor",
	["Configuration.Admins.EditRank.People.Content.PlayerTemplate._Name"] = "Visitor (@%s)",
	["Configuration.Admins.EditRank.People.Header.TLabel"] = "Visitors & team members",
	["Configuration.Admins.NewAdmin.BottomData.RankTitle"] = "Onboarding a New Administrative Tier",
	["Configuration.Admins.NewAdmin.BottomData.StepLabel"] = "Phase %s of %s",
	["Configuration.Admins.NewAdmin.Page1.Body"] = "The Rank Onboarding Suite empowers you to establish new administrative tiers and assign corresponding applications and personnel. Individuals you designate will immediately gain access to Administer, regardless of their current session.\nAccess is restricted to applications you are authorized for. Only personnel with explicit permissions may enable universal application access.",
	["Configuration.Admins.NewAdmin.Page1.Header"] = "Welcome to the Rank Onboarding Suite",
	["Configuration.Admins.NewAdmin.Start"] = "+ Initiate Administative Tier",
	["Configuration.Admins.NewAdmin.Page1.NextPage.LogoWordmark"] = "Proceed",
	["Configuration.Admins.NewAdmin.Page2.FilterHdr"] = "Define attributes for this tier to ensure precise identification. The tier's designation must be unique, professional, and no longer than 50 characters.",
	["Configuration.Admins.NewAdmin.Page2.LogoWordmark"] = "Designation",
	["Configuration.Admins.NewAdmin.Page2.PreviewText"] = "Draft: %s",
	["Configuration.Admins.NewAdmin.Page3.AddPane.AddGroup.LogoWordmark"] = "Community",
	["Configuration.Admins.NewAdmin.Page3.AddPane.AddUser.LogoWordmark"] = "Visitor",
	["Configuration.Admins.NewAdmin.Page3.LogoWordmark"] = "Personnel",
	["Configuration.Admins.NewAdmin.Page3.PageHdr"] = "Select personnel for this tier. Adjustments can be made subsequently.",
	["Configuration.Admins.NewAdmin.Page4.LogoWordmark"] = "Privileges",
	["Configuration.Admins.NewAdmin.Page4.PageHdr"] = "Assign applications and commands authorized for this tier.",
	["Configuration.Admins.NewAdmin.Page5.Body"] = "Success! Your administrative tier has been established. Assigned personnel now have immediate access. Future adjustments can be made by revisiting this interface and utilizing the configuration icon next to the tier's designation.",
	["Configuration.Apps.Content.Template.InstallDate"] = "Incepted and configured %s",
	["Configuration.Apps.Content.Template.Local.HeaderLabel"] = "Local applications shan't be modified.",
	["Configuration.Apps.Header.Desc"] = "In this page, you may view the applications you have requested installation to your experience.",
	["Configuration.Apps.MessageBox.Content"] = "Are you sure you would like to discord of this application's installation status within your experience? In order to re-require it, you must visit the Application Marketplace.",
	["Configuration.Apps.MessageBox.Header"] = "Discard of %s?",
	["Configuration.Apps.Options.Frame.HeaderLabel"] = "Tweak the configuration for this application",
	["Configuration.ErrorLog"] = "Exception",
	["Configuration.ErrorLog.ScrollingFrame.Template"] = "Warning",
	["Configuration.InfoPage.Credits.License"] = "The administration panel you are utilizing, Administer, happens to be made possible by many freely-licensed applications.",
	["Configuration.InfoPage.Credits.title"] = "Sources & Citations",
	["Configuration.InfoPage.Donate.Button"] = "Purchase %s wealth redistributable",
	["Configuration.InfoPage.Donate.Message"] = "Administer is actively pioneering next-generation operational resources and platform augmentations without transactional barriers. Our initiatives remain perpetually cost-neutral. We encourage strategic contribution if Administer delivers quantifiable enhancement to your digital enterprise or personnel efficiency. <b>10% of your contribution is redistributed to your experience! Contributions are distributed among Administer’s staff personnel network.</b>",
	["Configuration.InfoPage.Donate.title"] = "Capital Donations",
	["Configuration.InfoPage.VersionDetails.Update"] = "Verify",
	["Configuration.InfoPage.License"] = "Administer is architected and continuously evolved by Administer Software, a forward-thinking consortium of digital governance innovators, revolutionizing the Roblox administrative paradigm.\nAll public-facing deliverables are proudly aligned with free code sources with the GNU AGPL 3.0 license.\nUtilization, redistribution, and derivative creation of Administer, affiliated marketplace modules, and platform services are regulated under these corporate governance protocols.\nFor full access to these regulatory terms, please consult the repository on GitHub. To review the privacy compliance framework of AOS, refer to the dedicated DevForum message board.\n\nAdminister Software %s",
	["Configuration.InfoPage.License.About"] = "Free Code Utilization Terms",
	["Configuration.Marketplace.Header.Head"] = "Administer Program Storefront",
	["Configuration.Marketplace.Header.Plugins"] = "Applications",
	["Configuration.Marketplace.Header.Themes"] = "Interface Styling",
	["Configuration.Marketplace.Install.Created"] = "Incepted <b>%s</b>",
	["Configuration.Marketplace.Install.Updated"] = "Modified <b>%s</b>",
	["Configuration.Marketplace.Install.MetaServer"] = "You have seen this due to the fact that <b>You're subscribed to %s</b>",
	["Configuration.Marketplace.Install.MetaVersion"] = "Uses AppPlatform %s, which is %s by your Administer version.",
	["Configuration.Marketplace.MPFrozen.HeaderLabel"] = "Our apologies.",
	["Configuration.Marketplace.MPFrozen.Subheading1"] = "Sorry, but one or more app servers returned an error while processing that (code: %s, route %s). This may be a ban, a temporary ratelimit, or it may be unavailable. Please retry your request again soon. If you keep seeing this page please check the log and remove any defective app servers.",
	["Configuration.Marketplace.MainMarketplace.Content.Template.InstallLabel"] = "See other",
	["Configuration.Marketplace.Searchbar.Searchbar.Label"] = "Search for applications, interface styles, and plug-ins...",
	["Configuration.Marketplace.Searchbar.Searchbar.Popup.Label"] = "Insert text to execute an applet...",
	["Configuration.MenuBar.New.AInfoPage.HeaderLabel"] = "Information",
	["Configuration.Settings.CategoriesPage.BoolTemplate.Action.Label"] = "Revoke enablement",
	["Configuration.Apps.Description"] = "%s\n\n%s created by @%s from the instance %s",
	["Header.AppDrawer.HeaderLabel"] = "Applications",
	["Header.Mark.HeaderLabel"] = "<b>%s</b> · %s",
	["Home.Welcome"] = "<stroke color: \"rgb(0,0,0)\" transparency: \"0.85\" thickness: \".4\">Administer is loading...</stroke>",
	["Home.Widget1.Editing.Last"] = "Prior",
	["Home.Widget1.Editing.Next"] = "After current",
	["Import.Select.Header.Head"] = "Let's help ease the transition. Select what to import from where.",
	["NotFound.ErrorContent.TextLabel"] = "If you're looking for a menu, it's not here.",
	["NotFound.PageBroken.TextLabel"] = "This page did something that Roblox or Administer could not understand. It cannot continue loading.",
	["NotFound.PageBroken.sending"] = "This app (%s) had some form of runtime or Administer API error, which is fatal and would freeze something otherwise. Please report this error to the developer.",
	["Welcome.Initial.HeaderLabel"] = "Greetings and salutations, %s! %s",
	["widgets.uninstalled.Header"] = "This application menu appears to have been disabled.",
	["widgets.uninstalled.Body"] = "You may have disabled it's runtime execution or instructed Administer to discard of it. Please select a new application.",
	["widgets.notselected.Header"] = "Choose an application to be displayed.",
	["widgets.concurrent.Body"] = "To continue to edit application widgets, you msut first end your current session.",
	["widgets.welcome.Body"] = "To get started with widgets, just add an application to your collection and then inquire to see which applications support the Widget system.",
	["admincheck.OVERRIDE"] = "Located from within the administrator override",
	["admincheck.SUDO"] = "Sandbox mode (sudo) is enabled in your game settings or Variables script.",
	["admincheck.RANKINDEX"] = "Found in the ranks index",
	["admincheck.GROUP"] = "Data based on group rank",
	["admincheck.nil"] = "User not found in the admin index",
	["admin.RankDefault"] = "Administrator",
	["admin.NoRank"] = "Not an administrator",
	["admin.InOverride"] = "This user is in the override module, as such we don't have any information.",
	["admin.group.AllRanks"] = "Administrative Tier Directory",
	["admin.rankcreate.MessagingServiceFailed"] = "We made the rank fine, but failed to publish the event to tell other servers to check. Please try freeing up some MessagingService slots. {e}",
	["admin.rankcreate.Done"] = "This administrative tier has been incepted.",
	["admin.ranktype.System"] = "Experience Creator Overlord",
	["admin.ranktype.Protected"] = "Administrative Tiers that have been marked as inprotectable shan't be modified by anybody except for your expeirence admistrator.",
	["admin.ranktype.Protected.meta"] = "The Administrative Tier that you requested has been locked.",
	["admin.rankcreator.BadData"] = "You must provide a valid AdminRank!",
	["admin.rank.Default"] = "Added by System for first-time setup.",
	["admin.rankcreator.error.Generic"] = "We couldn't process that request right now, try again later.\\n\\n%s",
	["admin.rank.created"] = "Created from the rank editor.",
	["admin.rankcreator.corrupt.Detail"] = "Improper Rank Configuration Detected",
	["admin.rankcreator.corrupt.Body"] = "Administer has detected an error in the rank configuration. This is most likely due to an incomplete setup process and will result in Administer not loading out of Studio. Please re-initialize Administer.\n%s",
	["admin.ranks.superadmin.Detail"] = "Super Administrator",
	["admin.ranks.superadmin.Body"] = "Because this user is a Super Administrator, they'll be able to access any App (currently installed and installed in the future), manage + audit system functionality, and use restricted functionalities.",
	["admin.ranks.Others"] = "%s other entities...",
	["admin.ranks.members.one"] = "has been made a follower of this administrative tier",
	["admin.ranks.members.two"] = "are followers of this administrative tier",
	["admin.ranks.members.more"] = "%s others are members",
	["admin.ranks.protected.Detail"] = "Insufficient permissions",
	["admin.ranks.protected.Body"] = "Administrative Tiers that are protected shan't have modifications directly drafted..",
	["admin.rankcreator.failed.Detail"] = "Oops! That's not meant to happen",
	["admin.rankcreator.errors.filter.Body"] = "Failed to connect to Roblox's filtering service. This likely isn't an issue with your game, try again later.\n\n%s",
	["admin.rankcreator.errors.name.Len"] = "The primary identification for this Administrative Tier must induce more than 50 graphemes but one at minimum.",
	["admin.rankcreator.errors.name.Unique"] = "The primary identification for administrative tiers must be unique. Therefore, %s may not be used again.",
	["admin.rankcreator.errors.members.NotEnough"] = "Administrative tiers must have one entity at minimum, that being a community or visitor.",
	["admin.rankcreator.errors.members.TooMany"] = "The current maximum for administrative tier entities; communities or visitors, is %s..",
	["admin.rankcreator.errors.members.NoInput"] = "One or more of your members do not have a UserID assigned. All users must have an ID.",
	["generic.Success"] = "Task successfully completed with an expected result.",
	["generic.Unauthorized"] = "You lack proper permissions from your Administrative Tier.",
	["generic.error.CheckLogs"] = "It appears as if an exception has occurred. Please refer to your experience's scroll.",
	["generic.error.Generic"] = "It appears as if an exception has occurred.",
	["generic.error.TryAgainLater"] = "This action may not be taken at the moment, please re-attempt this action at a future date.",
	["generic.Please"] = "I beg",
	["generic.ThankYou"] = "I extend my greatest gratitude.",
	["generic.Created"] = "Incepted",
	["generic.Done"] = "Completed with an expected result",
	["generic.Updated"] = "Modification complete & changed",
	["generic.Install"] = "Add to experience library",
	["generic.Load"] = "Request content download",
	["generic.Loading"] = "This content is downloading.",
	["generic.Group"] = "Community",
	["generic.User"] = "Visitor",
	["generic.Yes"] = "Indeed",
	["generic.No"] = "Negatory",
	["generic.OK"] = "Accepted",
	["generic.UI"] = "User Interface",
	["generic.Cancel"] = "Disregard",
	["generic.Confirm"] = "Complete action",
	["generic.Download"] = "Pull content",
	["generic.Remove"] = "Discard of",
	["generic.Protected"] = "Elevated Security",
	["generic.admin"] = "Administrator",
	["generic.Forward"] = "Forward",
	["generic.Exit"] = "Disregard or Revoke",
	["generic.NotFound"] = "Record missing",
	["generic.Supported"] = "Valid support",
	["generic.Unsupported"] = "Invalid support",
	["generic.InvalidInput"] = "Input does not endure validation",
	["generic.UpToDate"] = "Synced with repository",
	["generic.OutOfDate"] = "Missing one or more changes",
	["generic.Registration"] = "Registration",
	["generic.Locally"] = "In your experience",
	["onboarding.Welcome"] = "%s has finished the loading stage! Your Administrative Tier is currently configured as a(n) %s. You may press %s to trigger Administer.",
	["onboarding.Head"] = "Greetings! Welcome to %s.",
	["apps.NoFetch"] = "Applications may not be fetched at the moment.",
	["apps.addserver.OK"] = "This AOS instance has been saved and registered as primary.",
	["apps.addserver.Failed"] = "Sorry, but something went wrong while assigning your primary instance. Please make note of the error in your game's logs and report it to the server maintainer",
	["apps.RunningSince"] = "Execution began at",
	["apps.BadServer"] = "This is not a valid Administer AOS instance.",
	["apps.Disable"] = "You can re-enable it from the \"Disabled Apps\" menu. The app may be able to continue running for this session but it will not be started in any new servers.",
	["apps.Remove"] = "This app will not start in any new servers but will continue running.",
	["apps.DupeDetected"] = "This application appears to already be in your experience's library; in order to re-initialzie it, it must be discarded.",
	["apps.Downloaded"] = "Successfully installed an app!",
	["apps.BadID"] = "Bad app ID! This app server is returning bad data.",
	["apps.server.Down"] = "Failed reaching out to the App Server. Perhaps it's offline?",
	["apps.server.NotAdminister"] = "This is not an AOS instance! Check logs for more info.",
	["apps.pulled.Bad"] = "Something went wrong with this module: %s",
	["apps.aos.Down"] = "AOS didn't return an OK status code",
	["apps.aos.InternalError"] = "Internal error",
	["apps.sensitivity.Context"] = "Utilizies the following %s Administer modules and Application Interfaces: %s",
	["apps.sensitivity.Sensitive"] = "highly sensitive",
	["apps.sensitivity.Insensitive"] = "negligable sensitivity",
	["apps.sensitivity.DoesNotUse"] = "Does not use %s application interfaces",
	["apps.installed.Body"] = "The application has been added to your experience's library. However, in order for frontends to be properly initialized, you must re-join the experience..",
	["apps.search.NoResults"] = "Sorry, but we couldn't find any results for that",
	["apps.search.FoundMatch"] = "We think you meant %s (%s% confidence), results for it may be included.",
	["apps.load.Fail"] = "Applications may not load currently.",
	["apps.load.Ratelimit"] = "Rate limit exceeded, please slow down",
	["apps.score.Detail"] = "Why is this application visible at the top position?",
	["apps.score.Body"] = "This application is currently at the top position due to the fact that the Administer Online Services instance that it originates from, %s ({}), has reported its weighted score to be %s, which is the highest among your installed servers.\nIf this is being done maliciously, remove the server.",
	["apps.aos.Out"] = "This server seems to be experiencing an outage. Try installing another in the meantime.",
	["apps.aos.Uninstall"] = "Uninstalling app servers may have unintended side effects. Please make sure you have at least one server installed.",
	["apps.aos.install.Detail"] = "Implement custom AOS instance",
	["apps.aos.install.Body"] = "Please enter your custom AOS server here with https:// and no trailing slash. It must be online and linked to your Mongo instance in order to be saved as primary.",
	["apps.aos.install.CantUninstall"] = "Cannot uninstall",
	["apps.aos.install.CantInstall.detail.NotEnoughServers"] = "To remove this server, you must first add a new one.",
	["apps.library.Detail"] = "%s apps, %s servers",
	["apps.published.Context"] = "Published by <b>%s</b>",
	["apps.published.Official"] = "This server is <b>trusted.</b>",
	["apps.published.ThirdParty"] = "This server is 3rd party and <b>untrusted.</b>",
	["mobile.body"] = "Mobile mode activated. Swipe inwards from the right edge of your screen to open Administer.",
	["mobile.detail"] = "Mobile mode",
	["error.NoPing"] = "Administer server ping failed, it seems your client may be incorrectly installed or the server is not executing properly. Please reinstall from source.",
	["error.StartupFailed"] = "Startup Failed",
	["error.LostDep"] = "Failed to boot client, missing required dependency (%s), please reinstall or roll back any recent changes",
	["error.LostDep.ClientFacingError"] = "Startup aborted, please make sure Administer is correctly installed. (failed dependency: %s)",
	["error.BootFail"] = "Boot failure",
	["update.Check"] = "Check for Updates",
	["updater.Failed.Body"] = "Update thread is frozen, please ensure AOS is responsive. Some services may be unavailable",
	["updater.Failed.Header"] = "Version check failed",
	["updater.failedcomm"] = "Unable to communicate with AOS.",
	["donations.Done"] = "Thank you for your support, %s! Your donation helps ensure future Administer updates for years to come ^^",
	["donations.DonationProcessed"] = "Your donation helps give back to the open-source community and will ensure: - lots of updates in the future - development funding for more apps - app server hosting Your donation means a lot to us! Make sure to show us this to claim your donor perks.",
	["donations.BadProxy"] = "The proxy may be offline or experiencing issues.",
	["updater.outofdate.Body"] = "%s is out of date. Please update your module.",
	["updater.outofdate.Header"] = "Version check complete",
	["filter.failed"] = "Failed to filter!",
	["setting.Admins.CheckTime.DisplayName"] = "Admin check time",
	["setting.Admins.CheckTime.Description"] = "How often to check if admins are still admins, and take away the panel if they are not.",
	["setting.Admins.AllowProtectedEdits.DisplayName"] = "Edit Protected Ranks",
	["setting.Admins.AllowProtectedEdits.Description"] = "Allows editing protected ranks (superadmin). Also allows the creation of them. Not recommended as you could potentially break your installation.",
	["setting.Interface.AnimationSpeed.DisplayName"] = "Animation Speed",
	["setting.Interface.AnimationSpeed.Description"] = "Controls the speed of client side animations. Higher is slower. Set to 0 to disable animations.",
	["setting.Interface.DisableUnofficialAppWarning.DisplayName"] = "Disable unofficial app warning",
	["setting.Interface.DisableUnofficialAppWarning.Description"] = "Hides the warning which says if a plugin is external. Useful for new people, otherwise just UI clutter.",
	["setting.Interface.EnableEditableImages.DisplayName"] = "Enable EditableImages",
	["setting.Interface.EnableEditableImages.Description"] = "Enables EditableImage features such as app card backgrounds, app card reflections, and other blurring effects. Does not disable \"image glow\". Currently nonfunctional outside of Studio due to Roblox limitations.",
	["setting.Interface.HomepageGreeting.DisplayName"] = "Homepage Greeting",
	["setting.Interface.HomepageGreeting.Description"] = "This is the text that displays after the \"Good morning, username!\" text on the homepage.",
	["setting.Interface.HomepageGreeting.Value"] = "Welcome to Administer!",
	["setting.Interface.PanelKeybind.DisplayName"] = "Panel Keybind",
	["setting.Interface.PanelKeybind.Description"] = "The key used to open the panel on keyboard-enabled devices.",
	["setting.Interface.EnableClickEffects.DisplayName"] = "Enable button effects",
	["setting.Interface.EnableClickEffects.Description"] = "Increases memory usage slightly, also makes noise.",
	["setting.Apps.LoadDelay.DisplayName"] = "LoadDelay",
	["setting.Apps.LoadDelay.Description"] = "Enables a 3s delay before loading any admins. Use this solve apps not initializing.",
	["setting.Apps.LoadDelay.Value"] = "None",
	["setting.Apps.FetchAppServers.DisplayName"] = "FetchAppServers",
	["setting.Apps.FetchAppServers.Description"] = "Disallows Administer from contacting any external app servers.",
	["setting.Apps.LoadApps.DisplayName"] = "LoadApps",
	["setting.Apps.LoadApps.Description"] = "Controls the ability for the server to load external apps. You shouldn't enable this unless you need a clean install or need to remove it.",
	["setting.Accessibility.ChatCommand.DisplayName"] = "ChatCommand",
	["setting.Accessibility.ChatCommand.Description"] = "Enables an /adm command to open Administer.",
	["setting.Accessibility.DisplayHours.DisplayName"] = "DisplayHours",
	["setting.Accessibility.DisplayHours.Description"] = "If enabled, hours will be displayed on all timestamps, may be ignored in some.",
	["setting.Accessibility.ShortNumberDecimals.DisplayName"] = "ShortNumberDecimals",
	["setting.Accessibility.ShortNumberDecimals.Description"] = "How many decimals to display next to short numbers. 2 -> 12.34k, 3 -> 12.345k, etc.",
	["setting.Accessibility.TopbarPlus.DisplayName"] = "TopbarPlus",
	["setting.Accessibility.TopbarPlus.Description"] = "Enables a Topbar+ section that you can launch Administer and apps from.",
	["setting.Accessibility.MobileOpenGesture.DisplayName"] = "MobileOpenGesture",
	["setting.Accessibility.MobileOpenGesture.Description"] = "Controls how big the mobile open gesture is from the right side. If the value is disabled you will need either the chat command or Topbar+ setting enabled.",
	["setting.Accessibility.MobileOpenGesture.Value"] = "regular",
	["setting.Miscellaneous.EditableImageRenderingDelay.DisplayName"] = "EditableImageRenderingDelay",
	["setting.Miscellaneous.EditableImageRenderingDelay.Description"] = "Enables a task.wait() call in EditableImage rendering calls. Can help reduce significant lag if present. WARNING: Makes client loading SIGNIFICANTLY slower.",
	["setting.Miscellaneous.NotificationCloseTimer.DisplayName"] = "NotificationCloseTimer",
	["setting.Miscellaneous.NotificationCloseTimer.Description"] = "Default waiting time to close the notification at. Can be overridden by notifications.",
	["setting.Miscellaneous.SandboxMode.DisplayName"] = "sudo",
	["setting.Miscellaneous.SandboxMode.Description"] = "(studio-only) Gives all people who join the game rank 1 (superadmin). Useful for setting up. Once your ranks are set up, you should turn this off. (also known as sudo)",
	["setting.Miscellaneous.Verbose.DisplayName"] = "Verbose",
	["setting.Miscellaneous.Verbose.Description"] = "Prints out general log data, useful for technical users or debugging. If enabled, it'll appear in the Log too.",
	["setting.category.Admins"] = "Admins",
	["setting.category.Interface"] = "Interface",
	["setting.category.Apps"] = "Apps",
	["setting.category.Accessibility"] = "Accessibility",
	["setting.category.Miscellaneous"] = "Miscellaneous",
	["setting.Locale.Swapped.Title"] = "Switched to English Corporate Speak",
	["setting.Locale.Swapped.Body"] = "This language exists for fun and is mostly a joke to mock Roblox, it isn't going to make sense and that's the point. Please don't report grammatical errors.",
	["reltime.Now"] = "Just now",
	["reltime.Day"] = "day ago",
	["reltime.Days"] = "days ago",
	["reltime.Week"] = "week ago",
	["reltime.Weeks"] = "weeks ago",
	["reltime.Month"] = "month ago",
	["reltime.Months"] = "months ago",
	["reltime.Minute"] = "minute ago",
	["reltime.Minutes"] = "minutes ago",
	["reltime.Hour"] = "hour ago",
	["reltime.Hours"] = "hours ago",
	["reltime.Year"] = "year ago",
	["reltime.Years"] = "years ago",
	["reltime.Format"] = "%s %s"
}
